core/bin/dust_api.rs

```diff
@@ -1641,6 +1641,250 @@ async fn data_sources_delete(
     }
 }
 
+// Databases
+
+#[derive(serde::Deserialize)]
+struct DatabasesRegisterPayload {
+    database_id: String,
+    name: String,
+}
+
+async fn databases_register(
+    extract::Path((project_id, data_source_id)): extract::Path<(i64, String)>,
+    extract::Json(payload): extract::Json<DatabasesRegisterPayload>,
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+    match state
+        .store
+        .register_database(
+            &project,
+            &data_source_id,
+            &payload.database_id,
+            &payload.name,
+        )
+        .await
+    {
+        Err(e) => error_response(
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "internal_server_error",
+            "Failed to register database",
+            Some(e),
+        ),
+        Ok(db) => (
+            StatusCode::OK,
+            Json(APIResponse {
+                error: None,
+                response: Some(json!({
+                    "database": db
+                })),
+            }),
+        ),
+    }
+}
+
+async fn databases_retrieve(
+    extract::Path((project_id, data_source_id, database_id)): extract::Path<(i64, String, String)>,
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+
+    match state
+        .store
+        .load_database(&project, &data_source_id, &database_id)
+        .await
+    {
+        Err(e) => error_response(
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "internal_server_error",
+            "Failed to retrieve database",
+            Some(e),
+        ),
+        Ok(db) => match db {
+            None => error_response(
+                StatusCode::NOT_FOUND,
+                "database_not_found",
+                &format!("No database found for id `{}`", database_id),
+                None,
+            ),
+            Some(db) => (
+                StatusCode::OK,
+                Json(APIResponse {
+                    error: None,
+                    response: Some(json!({
+                        "database": db
+                    })),
+                }),
+            ),
+        },
+    }
+}
+
+#[derive(serde::Deserialize)]
+struct DatabasesListQuery {
+    data_source_id: Option<String>,
```

Comment:
I think this should be a separate get no? what's the purpose of this query to limit the list to on database?

Reply:
Not sure I understand what you mean -- You can either get all databases in the project, or get all databases inside of a data source depending on if `data_source_id` is `None` or not. 
Which one of these do you feel we should remove (and why) ?

Reply:
> limit the list to on database?

It's to limit the list on one **data_source**, not one database